operator/src/main/java/oracle/kubernetes/operator/helpers/DomainStatusPatch.java

```diff
@@ -6,46 +6,110 @@
    * @param reason the reason, a camel-cased string with no spaces
    * @param message a text description of the new status; may include multiple lines
    */
-  public static void updateDomainStatus(Domain domain, String reason, String message) {
-    new DomainStatusPatch(domain).update(reason, message);
+  static Step createStep(Domain domain, String reason, String message) {
+    return new DomainStatusPatch(domain, reason, message);
+  }
+
+  /**
+   * Update the domain status in response to an unprocessable entity error. This may involve either replacing
+   * the current status or adding to it.
+   * @param domain the domain to update
+   * @param apiException the exception reporting an unprocessable entity
+   */
+  static Step createStep(Domain domain, ApiException apiException) {
+    UnprocessableEntityBuilder builder = UnprocessableEntityBuilder.fromException(apiException);
+    return createStep(domain, builder.getReason(), builder.getMessage());
+  }
+
+  /**
+   * Update the domain status synchronously. This may involve either replacing the current status or adding to it.
+   * @param domain the domain to update
+   * @param reason the reason, a camel-cased string with no spaces
+   * @param message a text description of the new status; may include multiple lines
+   */
+  public static void updateSynchronously(Domain domain, String reason, String message) {
+    new DomainStatusPatch(domain, reason, message).update();
   }
 
-  private DomainStatusPatch(Domain domain) {
-    this.domain = domain;
+  private DomainStatusPatch(Domain domain, String reason, String message) {
+    name = domain.getMetadata().getName();
+    namespace = domain.getMetadata().getNamespace();
+    patchBuilder = getPatchBuilder(domain, reason, message);
   }
 
-  private void update(String reason, String message) {
-    if (reason == null || message == null) return;
+  @Override
+  public NextAction apply(Packet packet) {
+    Step step = new CallBuilder().patchDomainAsync(name, namespace, getPatchBody(), createResponseStep());
+    return doNext(step, packet);
+  }
 
+  private DefaultResponseStep<Domain> createResponseStep() {
+    return new DefaultResponseStep<>(getNext());
+  }
+
+  private static JsonPatchBuilder getPatchBuilder(Domain domain, String reason, String message) {
+    JsonPatchBuilder patchBuilder = Json.createPatchBuilder();
+    if (domain.getStatus() == null) {
+      patchBuilder.add("/status", JsonValue.EMPTY_JSON_OBJECT);
+      patchBuilder.add("/status/reason", reason);
+      patchBuilder.add("/status/message", message);
+    } else {
+      setSubField(patchBuilder, "/status/reason", domain.getStatus().getReason(), reason);
+      setSubField(patchBuilder, "/status/message", domain.getStatus().getMessage(), message);
+    }
+    return patchBuilder;
+  }
+
+  private static void setSubField(JsonPatchBuilder patchBuilder, String path, String oldValue, String newValue) {
+    if (oldValue == null)
```

Comment:
braces

Reply:
You know I hate braces for one-line conditional bodies :)

At some point, we should really have a discussion on the advantages and disadvantages.

Reply:
BTW, I have found that it is precisely the "smallest possible change" that leads to code and architecture degradation. It is for that reason that the Agile style of development urges aggressive refactoring. It keeps the code clean and flexible. That necessarily demands extensive unit tests, of course.

Reply:
exactly, if we had that level of test coverage, then i would have a different opinion

Reply:
Overall, we have 67% line coverage with unit tests. I've been increasing the coverage as I work, and other developers have been doing so as well. The class that you're looking at has 96% line coverage; the only line not covered is a catch for an exception that can never be thrown (but Java requires it to be caught).

It's mostly the old code that has poor test coverage.